### DISCUSSION TITLE
Malformed database, now can't start server

### DISCUSSION THREAD

**Discussion by mhaimes** [dani-garcia/vaultwarden#1533] [Q&A]
### Subject of the issue

[2021-03-24 17:29:30.983][panic][ERROR] thread 'main' panicked at 'Failed to turn on WAL: DatabaseError(__Unknown, "f
ile is not a database")': src/db/mod.rs:275

### Deployment environment
running on a google f1-micro instance in a docker container with all the recommended settings. Suddenly, sometime over the last day or two I'm not sure, it looks like it decided my database is corrupted.. now I can't even start the server or get my information out of it. 

* bitwarden_rs version:

* Install method:
docker image
* Clients used: 
web vault, desktop, browser extensions, android, etc.
* Reverse proxy and version: 
latest version as of 3/24/2021
* MySQL/MariaDB or PostgreSQL version: 
it uses sqlite
* Other relevant details:

### Steps to reproduce

It happens when I try to start the bitwarden docker image, I assume the server tries to start and dies because it can't parse the daatabase.
### Expected behaviour

I expected, like normally, when i do a restart, I stop the docker image, grab the newest version, and start it again
### Actual behaviour

I am now in a state where I am completely unable to start the server
### Troubleshooting data

[2021-03-24 17:29:30.983][panic][ERROR] thread 'main' panicked at 'Failed to turn on WAL: DatabaseError(__Unknown, "f
ile is not a database")': src/db/mod.rs:275
   0: bitwarden_rs::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/panicking.rs:595:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/panicking.rs:497:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/sys_common/backtrace.rs:141:18
   4: rust_begin_unwind
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/panicking.rs:493:5
   5: core::panicking::panic_fmt
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/panicking.rs:92:14
   6: core::option::expect_none_failed
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/option.rs:1266:5
   7: bitwarden_rs::db::DbPool::from_config
   8: bitwarden_rs::util::retry_db
   9: bitwarden_rs::main
  10: std::sys_common::backtrace::__rust_begin_short_backtrace
  11: std::rt::lang_start::{{closure}}
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/ops/function.rs:259:13
      std::panicking::try::do_call
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/panicking.rs:379:40
      std::panicking::try
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/panicking.rs:343:19
      std::panic::catch_unwind
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/panic.rs:396:14
      std::rt::lang_start_internal
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/std/src/rt.rs:51:25
  13: main
  14: __libc_start_main
  15: _start

**Comment by BlackDex** [comment]
Seems like the storage you use is either full, or doesn't support Locking which is needed for WAL.

**Reply by mhaimes** [reply]
The storage definitely supports locking, it's the standard bitwarden_rs docker image, and it's defintely not full. it's been running fine for months. I restart the docker image once a night, which seemed to prevent a memory leak bug of some sort, but that's not what's at issue here.

**Reply by mhaimes** [reply]
Here's the output from df showing that there's still plenty of space left (the docker image folders are mapped to a subfolder of /):

root@bitwarden:/bw-data# df
Filesystem     1K-blocks    Used Available Use% Mounted on
udev              289424       0    289424   0% /dev
tmpfs              59724    1824     57900   4% /run
/dev/sda1       10126600 3709328   5883156  39% /
tmpfs             298604       0    298604   0% /dev/shm
tmpfs               5120       0      5120   0% /run/lock
tmpfs             298604       0    298604   0% /sys/fs/cgroup
/dev/sda15        126710    8034    118676   7% /boot/efi
tmpfs              59720       0     59720   0% /run/user/1000

**Comment by BlackDex** [comment]
Could you try to start a containers on an other system with the same settings and exact same database? And not somewhere in Google Cloud, but your own system.

**Reply by mhaimes** [reply]
I tried this. All that happened was the container still failed to start and this time the logfile was a bunch of binary garbage, no readable errors. Fresh Ubuntu vm, fresh everything, plenty of space..

Is there a way to get a mass export of everything from any of the clients? They obviously still have all the data, but whenever I use their export function I get like 3 passwords...

**Comment by jjlin** [comment]
Can the `sqlite3` CLI load the database? You can make a copy of the database and try some of the recovery options listed at https://stackoverflow.com/q/18259692. Since you haven't mentioned restoring from [backups](https://github.com/dani-garcia/bitwarden_rs/wiki/Backing-up-your-vault), I assume you don't have any. It probably goes without saying, but now you know why you need them.

Also, you shouldn't need to be restarting every night. I'd guess it's not a memory leak, but something like #1515.
